include/cell.tcc

```diff
@@ -979,37 +979,106 @@ bool mpm::Cell<Tdim>::assign_velocity_constraint(unsigned face_id, unsigned dir,
   return status;
 }
 
-//! Compute all face normals 2d
+//! Apply velocity constraints
+template <unsigned Tdim>
+void mpm::Cell<Tdim>::apply_velocity_constraints() {
+
+  // Assum phase = 0 - TODO: make it general
+  const unsigned phase = 0;
+
+  // Copy nodal velocities and accelerations
+  for (unsigned i = 0; i < this->nfunctions(); ++i) {
+    constrained_nodal_velocity_.insert(
+        std::make_pair<unsigned, Eigen::VectorXd>(i,
+                                                  nodes_[i]->velocity(phase)));
+    constrained_nodal_acceleration_.insert(
+        std::make_pair<unsigned, Eigen::VectorXd>(
+            i, nodes_[i]->acceleration(phase)));
+  }
+
+  // Set velocity constraint
+  for (auto iterator = this->velocity_constraints_.begin();
+       iterator != this->velocity_constraints_.end(); ++iterator) {
+    // Get face id
+    const auto face_id = iterator->first;
+    // Loop through all the velocity constraints
+    for (const auto& constraint : this->velocity_constraints_[face_id]) {
+      // Direction value in the constraint (0, Dim * Nphases)
+      const auto dir = constraint.first;
+      // Direction: dir % Tdim (modulus)
+      const auto direction = static_cast<unsigned>(dir % Tdim);
+      // Phase: Integer value of division (dir / Tdim)
+      // const auto phase = static_cast<unsigned>(dir / Tdim);
+
+      // Get inverse rotation matrix
+      Eigen::Matrix<double, Tdim, Tdim> rotation_matrix =
+          geometry_->rotation_matrix(geometry_->euler_angles_cartesian(
+              this->new_coordinate_axes_[face_id]));
+
+      // Get the nodes of the face
+      const Eigen::VectorXi node_indices = element_->face_indices(face_id);
+
+      // Apply it to the nodes
```

Comment:
I can't see it being applied to any node here! There is no access to a single nodal pointer. Am I missing something?